phaser/phase.py

```diff
@@ -1,13 +1,153 @@
+        :param row: What row of the data this occurred in
+        :return: Nothing
+        """
+        if isinstance(exc, DropRowException):
+            self.context.add_dropped_row(step, row, exc.message)
+        elif isinstance(exc, WarningException):
+            self.context.add_warning(step, row, exc.message)
+        else:
+            e_name = exc.__class__.__name__
+            e_message = str(exc)
+            message = f"{e_name} raised ({e_message})" if e_message else f"{e_name} raised."
+            logger.debug(f"Unknown exception handled in executing steps ({message}")
+
+            match self.error_policy:
+                case Pipeline.ON_ERROR_COLLECT:
+                    self.context.add_error(step, row, message)
+                case Pipeline.ON_ERROR_WARN:
+                    self.context.add_warning(step, row, message)
+                case Pipeline.ON_ERROR_DROP_ROW:
+                    self.context.add_dropped_row(step, row, message)
+                case Pipeline.ON_ERROR_STOP_NOW:
+                    self.context.add_error(step, row, message)
+                    raise exc
+                case _:
+                    raise PipelineErrorException(f"Unknown error policy '{self.error_policy}'") from exc
+
+    def report_errors_and_warnings(self):
+        """ In next iteration, we should probably move the generation of error info to stdout to other locations.
+        For CLI operation we want to report errors to the CLI, but for unsupervised operation these should go
+        to logs.  Python logging does allow users of a library to send log messages to more than one place while
+        customizing log level desired, and we could have drop-row messages as info and warning as warn level so
+        these fit very nicely into the standard levels allowing familiar customization.  """
+        # LMDTODO: How are one phases's errors and warnings kept separate from another phase's?
+        for row_num, info in self.context.dropped_rows.items():
+            print(f"DROPPED row: {row_num}, message: '{info['message']}'")
+        # Unlike errors and dropped rows, there can be multiple warnings per row
+        for row_num, warnings in self.context.warnings.items():
+            for warning in warnings:
+                print(f"WARNING row: {row_num}, message: '{warning['message']}'")
+        for row_num, error in self.context.errors.items():
+            print(f"ERROR row: {row_num}, message: '{error['message']}'")
+
+
+class DataFramePhase(PhaseBase):
+    def __init__(self, name, context=None, error_policy=None):
+        super().__init__(name, context=context, error_policy=error_policy)
+        self.df_data = None
+
+    def run(self, source, destination):
+        self.load(source)
+        self.df_data = self.df_transform(self.df_data)
+        self.row_data = self.df_data.to_dict('records')
+        if self.context.has_errors():
+            self.report_errors_and_warnings()
```

Comment:
I'd move the call to `self.report_errors_and_warnings()` to above the `if/else` statement instead of repeating them.